torch/csrc/jit/constants.cpp

```diff
@@ -90,9 +90,20 @@ c10::optional<Value*> tryInsertConstant(
     n->setScope(*scope);
   if (result_type) {
     auto inferred_type = n->output()->type();
-    // Retain more type information in case of tensor constant
-    if (!(inferred_type->isSubtypeOf(TensorType::get()) &&
-          result_type->isSubtypeOf(inferred_type))) {
+
+    if (inferred_type->isSubtypeOf(NoneType::get()) &&
+        !inferred_type->isSubtypeOf(result_type)) {
+      // None doesn't subtype Optional, but an Optional can be None, so handle
+      // that here
+      if (result_type->kind() == TypeKind::OptionalType) {
+        n->output()->setType(result_type);
+      } else {
+        // Implicitly wrap non-optionals
+        n->output()->setType(OptionalType::create(result_type));
```

Comment:
How can this branch ever be reached ?  `inferred_type->isSubtypeOf(NoneType::get()) && !inferred_type->isSubtypeOf(result_type))` means that result_type has to be an Optional.

Reply:
This is in there to maintain calls like `torch.jit.annotate(Tensor, None)` where the type hint is `Tensor`

Reply:
I think this should be handled in the compiler and not at this callsite. InsertConstant is a common API and this usage would be a bug at any of the other callsites

Reply:
It's not a bug though, it's something we support currently. Plus keeping it here lets us have all usages in 1 place so we can more easily delete it later if we go that route

Reply:
It is a bug. In any other case having a type of Tensor for a value of type None is a bug